modules/consensus/synchronize.go

```diff
@@ -45,121 +42,121 @@ func (g *Gateway) threadedResynchronize() {
 // such transmissions may be required to fully synchronize.
 //
 // TODO: don't run two Synchronize threads at the same time
-func (g *Gateway) Synchronize(peer modules.NetAddress) error {
-	g.log.Println("INFO: synchronizing to", peer)
-	for {
+func (s *State) Synchronize(peer modules.NetAddress) error {
+	// loop until there are no more blocks available
+	moreAvailable := true
+
+	for moreAvailable {
+		// get blockIDs to send
+		id := s.mu.RLock()
+		history := s.blockHistory()
+		s.mu.RUnlock(id)
+
+		// perform RPC
 		var newBlocks []types.Block
-		newBlocks, moreAvailable, err := g.requestBlocks(peer)
+		err := s.gateway.RPC(peer, "SendBlocks", func(conn modules.PeerConn) error {
+			err := encoding.WriteObject(conn, history)
+			if err != nil {
+				return err
+			}
+			err = encoding.ReadObject(conn, &newBlocks, MaxCatchUpBlocks*types.BlockSizeLimit)
+			if err != nil {
+				return err
+			}
+			return encoding.ReadObject(conn, &moreAvailable, 1)
+		})
 		if err != nil {
-			g.log.Printf("ERR: synchronization to %v failed: %v\n", peer, err)
 			return err
 		}
-		g.log.Printf("INFO: %v sent us %v blocks\n", peer, len(newBlocks))
+
+		// integrate received blocks
 		for _, block := range newBlocks {
-			acceptErr := g.state.AcceptBlock(block)
+			// TODO: don't Broadcast these blocks
+			acceptErr := s.AcceptBlock(block)
 			if acceptErr != nil {
 				// TODO: If the error is a FutureTimestampErr, need to wait before trying the
 				// block again.
-				g.log.Printf("WARN: state rejected a block from %v: %v\n", peer, acceptErr)
 			}
 		}
-
-		// loop until there are no more blocks available
-		if !moreAvailable {
-			break
-		}
 	}
-	g.log.Printf("INFO: synchronization to %v complete\n", peer)
 	return nil
 }
 
-// sendBlocks returns a sequential set of blocks based on the 32 input block
+// SendBlocks returns a sequential set of blocks based on the 32 input block
 // IDs. The most recent known ID is used as the starting point, and up to
 // 'MaxCatchUpBlocks' from that BlockHeight onwards are returned. It also
 // sends a boolean indicating whether more blocks are available.
-func (g *Gateway) sendBlocks(conn modules.NetConn) (err error) {
+func (s *State) SendBlocks(conn modules.PeerConn) error {
```

Comment:
hmm. I guess the state could register its RPCs in `New`. It's convenient to have them all registered in the same place, but the gateway breaks that convention anyway...